File: src/main/java/com/hazza/LeetCodeByJava/SegmentTwo/No85MaximalRectangle.java
package com.hazza.LeetCodeByJava.SegmentTwo;

import java.util.Arrays;
import java.util.Stack;

/**
 * Created with IntelliJ IDEA.
 * Description:
 * Author: HazzaCheng
 * Contact: hazzacheng@gmail.com
 * Date: 2017-09-13
 * Time: 5:21 PM
 */
public class No85MaximalRectangle {

    // solution 1
    public int maximalRectangle(char[][] matrix) {
        if (matrix == null || matrix.length == 0|| matrix[0].length == 0) return 0;
        int rowLen = matrix.length, colLen = matrix[0].length;
        int[] left = new int[colLen];
        int[] right = new int[colLen];
        int[] height = new int[colLen];
        Arrays.fill(right, colLen);
        int maxArea = 0;
        for (int i = 0; i < rowLen; i++) {
            int curLeft = 0, curRight = colLen;
            for (int j = 0; j < colLen; j++) {
                if (matrix[i][j] == '1') ++height[j];
                else height[j] = 0;
            }
            for (int j = 0; j < colLen; j++) {
                if (matrix[i][j] == '1') left[j] = Math.max(left[j], curLeft);
                else {
                    left[j] = 0;
                    curLeft = j + 1;
                }
            }
            for (int j = colLen - 1; j >= 0; j--) {
                if (matrix[i][j] == '1') right[j] = Math.min(right[j], curRight);
                else {
                    right[j] = colLen;
                    curRight = j;
                }
            }
            for (int j = 0; j < colLen; j++) {
                maxArea = Math.max(maxArea, (right[j] - left[j]) * height[j]);
            }
        }
        return maxArea;
    }

    // solution 2
    public int maximalRectangle1(char[][] matrix) {
        if (matrix == null || matrix.length == 0|| matrix[0].length == 0) return 0;
        int rowLen = matrix.length, colLen = matrix[0].length;
        int[] heights = new int[colLen + 1];
        heights[colLen] = 0;
        int maxArea = 0;

        for (int i = 0; i < rowLen; i++) {
            Stack<Integer> stack = new Stack<>();
            for (int j = 0; j < colLen + 1; j++) {
                if (j < colLen) {
                    if (matrix[i][j] == '1') heights[j]++;
                    else heights[j] = 0;
                }
                if (stack.isEmpty() || heights[stack.peek()] <= heights[j]) stack.push(j);
                else {
                    while (!stack.isEmpty() && heights[stack.peek()] > heights[j]) {
                        int top = stack.pop();
                        maxArea = Math.max(maxArea,
                                    heights[top] * (stack.isEmpty() ? j : (j - stack.peek() - 1)));
                    }
                    stack.push(j);
                }
            }
        }

        return maxArea;
    }
}

File: src/main/python/com/hazza/leetcode_by_python/dp/No85_Maximal_Rectangle.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019-07-23
# https://leetcode.com/problems/maximal-rectangle/
from typing import List


class Solution:
    # dp
    def maximalRectangle1(self, matrix: List[List[str]]) -> int:
        if not matrix:
            return 0
        n = len(matrix[0])
        heights, left, right = [0] * (n + 1), [0] * (n + 1), [n] * (n + 1)
        res = 0

        for row in matrix:
            curLeft, curRight = 0, n
            for i in range(n):
                if row[i] == '1':
                    heights[i] += 1
                    left[i] = max(left[i], curLeft)
                else:
                    heights[i] = 0
                    curLeft = i + 1
                    left[i] = 0
            for i in range(n)[::-1]:
                if row[i] == '1':
                    right[i] = min(right[i], curRight)
                else:
                    curRight = i
                    right[i] = n
            for i in range(n):
                res = max(res, (right[i] - left[i]) * heights[i])

        return res

    # stack
    def maximalRectangle2(self, matrix: List[List[str]]) -> int:
        if not matrix:
            return 0
        n = len(matrix[0])
        heights = [0] * (n + 1)
        res = 0

        for row in matrix:
            stack = [-1]
            for i in range(n + 1):
                if i < n:
                    heights[i] = heights[i] + 1 if row[i] == '1' else 0
                while heights[i] < heights[stack[-1]]:
                    h = heights[stack.pop()]
                    w = i - stack[-1] - 1
                    res = max(res, h * w)
                stack.append(i)

        return res


if __name__ == '__main__':
    solution = Solution()

    matrix = [
        ["1", "0", "1", "0", "0"],
        ["1", "0", "1", "1", "1"],
        ["1", "1", "1", "1", "1"],
        ["1", "0", "0", "1", "0"]
    ]
    assert solution.maximalRectangle1(matrix) == 6
    assert solution.maximalRectangle2(matrix) == 6
